### PR TITLE
[move IR] Migrated to abilities

### PR DESCRIPTION
**Pull request by tnowacki**
- Added abilities to IR
- Made drop, store, and key a keyword. This can be improved for the source language
- Updated various tests. New tests to follow later

## Motivation

- Bringing abilities to the IR. Source language work and tests still to follow 

## Test Plan

- updated existing tests

### CODE REVIEW

**Review by bob-wilson — COMMENTED** [review]
This all looks fine.... except I really do not like the syntax. Sure, this is just the IR compiler and we could do something different in the source language, but at least for the basic type definitions it would be strange if they are very different. So, I'd like to have more discussion about the syntax before we move forward with this. I have two concerns: (1) it would be nice to avoid adding reserved keywords for things like "copy" and "drop", and (2) I find it hard to read type definitions that can start with so many different keywords. For the second point, you had made a comment earlier that we could follow the a convention of always putting the modifiers on the preceding line, but that's not what I see happening here, and it also seems like that would not be very readable, especially for short modifiers like "key struct". I had mentioned in a Google doc that I'd like to consider putting the new abilities in parenthesis like we're doing for visibility, e.g., "struct(key)" or "struct(drop copy)". With that you'd be able to easily scan for things starting with "struct" to find type definitions, and I think we would also be able to avoid making those into reserved keywords.

**Comment by tnowacki** [comment]
The syntax is moot for making these things keywords or not. I have a PR in progress for the source language that makes these not keywords; originally, I had thought it would be a bit tricky and avoided it for the IR. But it was easier than I expected so I can bring it down into this PR. 

I agree that I'm not super pleased with the syntax here, but I currently like it more than any other suggestion I have seen. That being said `struct(key)` or something similar isn't that bad in comparison to some others :) 

It's a bit hard to see in the tests but the syntax I have been going with is if it fits all on one line, do that
```
copy struct Foo { ... }
```
But if it is a multiline one, I have been putting it above the struct decl
```
key store
struct Foo {
    f1: u64,
    f2: u128,
    my_address: address,
    bar: Bar,
}
```
I think this helps a bit

In general, I really hate a bunch of modifiers. And have generally scoffed at languages that have a bunch. I will give alternatives a deep consideration, but so far I think this is okay (and just okay)

**Comment by sblackshear** [comment]
> I had mentioned in a Google doc that I'd like to consider putting the new abilities in parenthesis like we're doing for visibility, e.g., "struct(key)" or "struct(drop copy)

I have a minor issue with this proposal. In our other use of paren syntax (`public(<modifier>)`), the parens can only contain one modifier. But on structs, there can be several. Using this syntax for structs as well might make someone think that it's valid to write (e.g.) `public(script friend)`.

For this reason, I have a slight preference for the approach in this PR, although I agree with the importance of visually identifying declarations + think we should adopt the multiline convention to address this.

**Comment by tnowacki** [comment]
I think if I had to summarize in a few points why I like this over other syntax choices I have seen thus far
- It is non intrusive, as in you just add the modifiers and don't need some other bit of syntax
- It is flexible to different formatting styles, I like that some decisions about where to place these are very flexible for a future auto formatter 
- They are distinct from type parameter constraints, but share the same keywords so things line up. T: copy, requires a copy struct. This is just a note about defaults really, and why I wouldn't like something like `!copy struct Foo` where you are removing abilities or something
 
My hope is that these points help us navigate other possible designs and maybe even point to a better one

**Comment by bob-wilson** [comment]
> I have a minor issue with this proposal. In our other use of paren syntax (`public(<modifier>)`), the parens can only contain one modifier. But on structs, there can be several. Using this syntax for structs as well might make someone think that it's valid to write (e.g.) `public(script friend)`.

I look at that as a semantic restriction. From a syntactic perspective, it should be fine to have multiple modifiers on the "public(...)" visibility, and perhaps someday we might even find a reason to allow that (with different modifiers than what we have now).

I'm not necessarily attached to using the same parenthesis notation, but I'd much prefer a syntax that consistently starts with "struct" so you more easily find the types. One of the things I dislike about C/C++ is the way that function return types come before the names -- you can adopt conventions like putting the return type on a separate line, but I still find it much harder to read.

**Comment by tnowacki** [comment]
>  I'd much prefer a syntax that consistently starts with "struct" so you more easily find the types.

Can you go into more detail of what you mean by this? Not sure I follow exactly. But my thoughts on it if I take this at face value:
- I think if it is on one line there probably aren't many issues visually scanning for the type
- If it is a big, multi-line block, visually scanning is helped when the abilities are separate and above the struct decl
- Putting anything between `struct` and the struct's name will hurt tool based (non-visual) location of the type as you can no longer grep for `struct MyStructName`

**Comment by tnowacki** [comment]
> My point about putting "struct" first is just that when scanning code, especially code that other people have written, it should be as clear as possible what a particular blob of text is about. If the first thing you see is "struct....", then it is immediately clear that this is a type definition of some kind. The various abilities are secondary to that -- they are modifiers to describe what kind of struct it is, and having some syntactic clues to indicate that relationship is helpful.

I would keep in mind that this doesn't hold up directly when we eventually add public/private structs. As people will have to look past the `public` modifier and see if it is a struct/function. As happens in Rust, your eye gets trained to look for the pattern and you sort of ignore the visibility or look past it. 
I think a few things will happen here with abilities. 
- In the multi-line case, it is a non-issue. 
- In the single-line case, people will sort of learn to look past it much like the do for visibility
- Additionally for the single-line case, these keywords only appear by themselves for structs (and maybe enums when they eventually arrive) and so it will be easy to see "copy" or "drop" or etc and subconsciously know that it is a datatype declaration. 

> I'm not a fan of putting the abilities after the type name. The only other idea I can think of would be to somehow wrap the abilities in some syntax to try to convey that they are modifiers. I'm not sure I even like that idea but I'll throw it out in case it sparks ideas: e.g., "(copy drop) struct foo" or "[copy, drop] struct foo", etc.

I have tried some ideas like that. Without a keyword, () or [] doesn't add much to me. With a modifier, e.g. `abilities[copy, drop]` it stands out more but I almost don't like that it does. It makes it feel somehow less important or more compiler-auto-generated-y to me. In other words, it makes it feel less like a core aspect of the language.
And for reasons discussed above, it doesn't change the readability of it (at least for me)

**Comment by bob-wilson** [comment]
I don't like the "abilities" keyword. I do think the brackets (or parens) make a significant different in readability to my eyes (which admittedly have not yet been trained to recognize these patterns).

public [copy, drop] struct foo

is much easier for me to read than:

public copy drop struct foo

**Comment by tnowacki** [comment]
Ah, putting it after the visibility modifier does help a bit with brackets. 
I had been envisioning `copy drop public struct Foo`

Let me think about this a bit more, it might work out nicely. 

The thing I don't like about initially, is that we might want to allow `[] struct Foo` in the grammar, which is a bit weird to me, but maybe it is okay...

**Comment by tnowacki** [comment]
I'm not sure I really like brackets in that position, because at some point I probably want `#[]` for attributes, which could get confusing. I want to make sure these modifiers feel like a part of the type system, not some compiler driven attribute or modifier
---
That being said, we should really consider the long term appearance of these modifiers. We will likely be adding public and public(friend) for structs. And at some point I would like positional structs as a source language feature. 
So all of the cases to consider
```
struct Foo {}
struct Foo { f1: t1 }
struct Foo<T: X> { f1: t1 }
struct Foo()
struct Foo(t1)
struct Foo<T: X>(t1)
```
And with visibility
```
public struct Foo {}
public struct Foo { f1: t1 }
public struct Foo<T: X> { f1: t1 }
public struct Foo()
public struct Foo(t1)
public struct Foo<T: X>(t1)
public(friend) struct Foo {}
public(friend) struct Foo { f1: t1 }
public(friend) struct Foo<T: X> { f1: t1 }
public(friend) struct Foo()
public(friend) struct Foo(t1)
public(friend) struct Foo<T: X>(t1)
```

So applying different strategies to all of these examples
```
copy drop struct Foo {}
copy drop struct Foo { f1: t1 }
copy drop struct Foo<T: X> { f1: t1 }
copy drop struct Foo()
copy drop struct Foo(t1)
copy drop struct Foo<T: X>(t1)

struct Foo has copy, drop  {}
struct Foo has copy, drop  { f1: t1 }
struct Foo<T: X> has copy, drop  { f1: t1 }
struct Foo has copy, drop ()
struct Foo has copy, drop (t1)
struct Foo<T: X> has copy, drop (t1)

struct Foo[copy, drop] {}
struct Foo[copy, drop] { f1: t1 }
struct Foo<T: X>[copy, drop] { f1: t1 }
struct Foo[copy, drop]()
struct Foo[copy, drop](t1)
struct Foo<T: X>[copy, drop](t1)

struct Foo: copy + drop  {}
struct Foo: copy + drop  { f1: t1 }
struct Foo<T: X>: copy + drop  { f1: t1 }
struct Foo: copy + drop ()
struct Foo: copy + drop (t1)
struct Foo<T: X>: copy + drop (t1)

struct(copy drop) Foo {}
struct(copy drop) Foo { f1: t1 }
struct(copy drop) Foo<T: X> { f1: t1 }
struct(copy drop) Foo()
struct(copy drop) Foo(t1)
struct(copy drop) Foo<T: X>(t1)

copy drop public struct Foo {}
copy drop public struct Foo { f1: t1 }
copy drop public struct Foo<T: X> { f1: t1 }
copy drop public struct Foo()
copy drop public struct Foo(t1)
copy drop public struct Foo<T: X>(t1)

public struct Foo has copy, drop  {}
public struct Foo has copy, drop  { f1: t1 }
public struct Foo<T: X> has copy, drop  { f1: t1 }
public struct Foo has copy, drop ()
public struct Foo has copy, drop (t1)
public struct Foo<T: X> has copy, drop (t1)

public struct Foo[copy, drop] {}
public struct Foo[copy, drop] { f1: t1 }
public struct Foo<T: X>[copy, drop] { f1: t1 }
public struct Foo[copy, drop]()
public struct Foo[copy, drop](t1)
public struct Foo<T: X>[copy, drop](t1)

public struct Foo: copy + drop  {}
public struct Foo: copy + drop  { f1: t1 }
public struct Foo<T: X>: copy + drop  { f1: t1 }
public struct Foo: copy + drop ()
public struct Foo: copy + drop (t1)
public struct Foo<T: X>: copy + drop (t1)

public struct(copy drop) Foo {}
public struct(copy drop) Foo { f1: t1 }
public struct(copy drop) Foo<T: X> { f1: t1 }
public struct(copy drop) Foo()
public struct(copy drop) Foo(t1)
public struct(copy drop) Foo<T: X>(t1)

copy drop public(friend) struct Foo {}
copy drop public(friend) struct Foo { f1: t1 }
copy drop public(friend) struct Foo<T: X> { f1: t1 }
copy drop public(friend) struct Foo()
copy drop public(friend) struct Foo(t1)
copy drop public(friend) struct Foo<T: X>(t1)

public(friend) struct Foo has copy, drop  {}
public(friend) struct Foo has copy, drop  { f1: t1 }
public(friend) struct Foo<T: X> has copy, drop  { f1: t1 }
public(friend) struct Foo has copy, drop ()
public(friend) struct Foo has copy, drop (t1)
public(friend) struct Foo<T: X> has copy, drop (t1)

public(friend) struct Foo[copy, drop] {}
public(friend) struct Foo[copy, drop] { f1: t1 }
public(friend) struct Foo<T: X>[copy, drop] { f1: t1 }
public(friend) struct Foo[copy, drop]()
public(friend) struct Foo[copy, drop](t1)
public(friend) struct Foo<T: X>[copy, drop](t1)

public(friend) struct Foo: copy + drop  {}
public(friend) struct Foo: copy + drop  { f1: t1 }
public(friend) struct Foo<T: X>: copy + drop  { f1: t1 }
public(friend) struct Foo: copy + drop ()
public(friend) struct Foo: copy + drop (t1)
public(friend) struct Foo<T: X>: copy + drop (t1)

public(friend) struct(copy drop) Foo {}
public(friend) struct(copy drop) Foo { f1: t1 }
public(friend) struct(copy drop) Foo<T: X> { f1: t1 }
public(friend) struct(copy drop) Foo()
public(friend) struct(copy drop) Foo(t1)
public(friend) struct(copy drop) Foo<T: X>(t1)
```

**Comment by tnowacki** [comment]
Roping him in, @tzakian likes `struct(ability)` and `struct Foo<T:X>[ability]`

**Comment by metajack** [comment]
Another variant is you could just use attributes directly.

```
#[copy, drop]
public struct Foo {
  ...
  }
```

thought that feels kind of weird since annotations are usually extra things. Rust does use attributes for deriving Copy and Clone which are semi-related.

I suppose I'm still partial to the `: copy + drop` syntax (though not necessarily exactly that puncutation), with `[copy, drop]` after the type spot as the next choice.

`public(friend) struct(copy drop)` starts to get really tedious to mentally parse for me. Note that unlike visibility modifiers which should be rare-ish, these abilities will be on most types. That makes me wonder if we'll ever have alias keywords for the popular combos, like `all` for `copy, drop, key, store`, etc.

**Comment by tnowacki** [comment]
Recent fixup switches to the `has` syntax

File: language/compiler/ir-to-bytecode/syntax/src/lexer.rs [hunk 1]
```diff
@@ -434,7 +432,7 @@ fn get_name_token(name: &str) -> Tok {
         "old" => Tok::Old,
         "public" => Tok::Public,
         "requires" => Tok::Requires,
-        "resource" => Tok::Resource,
+        "copy" => Tok::Copy,
```
Comment (bob-wilson) [inline language/compiler/ir-to-bytecode/syntax/src/lexer.rs:435]: I don't know how "copyable" got added at the end, but this was supposed to be in alphabetic order to make it easier to find stuff. Can you move this to the right place?

File: language/move-ir/types/src/ast.rs [hunk 2]
```diff
@@ -301,8 +298,8 @@ pub struct FunctionSignature {
     pub formals: Vec<(Var, Type)>,
     /// Optional return types
     pub return_type: Vec<Type>,
-    /// Possibly-empty list of (TypeVar, Kind) pairs.s.
-    pub type_formals: Vec<(TypeVar, Kind)>,
+    /// Possibly-empty list of type paramters and their constraints
```
Comment (bob-wilson) [inline language/move-ir/types/src/ast.rs:301]: typo: "paramters"

**Review by bob-wilson — APPROVED** [review]
Thanks for the syntax changes! I like this much better. A few small nits but otherwise this is good to go.

File: language/compiler/ir-to-bytecode/syntax/src/lexer.rs (hunk 1 above)
Comment (tnowacki) [reply language/compiler/ir-to-bytecode/syntax/src/lexer.rs:435]: Ah probably a bad copy paste or something. I don't know. I'll fix it

**Pull request closed without merging**